File: 62 Unique Paths/62 LL.cpp
#include <iostream>
using namespace std;

/*
https://leetcode.com/problems/unique-paths/

A robot is located at the top-left corner of a m x n grid (marked 'Start' in the diagram below).

The robot can only move either down or right at any point in time.
The robot is trying to reach the bottom-right corner of the grid (marked 'Finish' in the diagram below).

How many possible unique paths are there?
*/

class Solution
{
public:
	int uniquePaths(int m, int n)
	{
		int c = min(m - 1 , n - 1);
		long long int result = (c < 0) ? 0 : 1;
		for (int i = 0; i < c; ++i)
		{
			// attention overflow may occur
			result = result * (m + n - 2 - i) / (i + 1);
		}
		return (int)result;
	}
};

/*
idea: pick (m - 1) right moves from (m - 1 + n - 1) total moves
complexity: Time O(N)
*/

int main(void)
{
	Solution engine;
	cout << engine.uniquePaths(3, 7) << '\n';
	return 0;
}

File: 62 Unique Paths/62 LL.py
class Solution(object):
	def uniquePaths(self, m, n):
		m -= 1
		n -= 1
		total = m + n
		select = min(m, n)
		result = 1 if select >= 0 else 0
		for i in xrange(select):
			result = result * (total - i) / (i + 1)
		return result

engine = Solution()
print engine.uniquePaths(2, 2)